internal/impl/redis/script_processor.go

```diff
@@ -0,0 +1,255 @@
+			Description("The time to wait before consecutive retry attempts.").
+			Default("500ms").
+			Advanced()).
+		Example("Running a script",
+			`The following example will use a script execution to get next element from a sorted set and set its score with timestamp unix nano value.`,
+			`
+pipeline:
+  processors:
+    - redis_script:
+        url: TODO
+        script: |
+          local value = redis.call("ZRANGE", KEYS[1], '0', '0')
+
+          if next(elements) == nil then
+            return ''
+          end
+
+          redis.call("ZADD", "XX", KEYS[1], ARGV[1], value)
+
+          return value
+        keys_mapping: 'root = [ meta("key") ]'
+        args_mapping: 'root = [ timestamp_unix_nano() ]'
+`)
+}
+
+func init() {
+	err := service.RegisterBatchProcessor(
+		"redis_script", redisScriptProcConfig(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.BatchProcessor, error) {
+			return newRedisScriptProcFromConfig(conf, mgr)
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type redisScriptProc struct {
+	log *service.Logger
+
+	script      *redis.Script
+	argsMapping *bloblang.Executor
+	keysMapping *bloblang.Executor
+
+	client      redis.UniversalClient
+	retries     int
+	retryPeriod time.Duration
+}
+
+func newRedisScriptProcFromConfig(conf *service.ParsedConfig, res *service.Resources) (*redisScriptProc, error) {
+	client, err := getClient(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	retries, err := conf.FieldInt("retries")
+	if err != nil {
+		return nil, err
+	}
+
+	retryPeriod, err := conf.FieldDuration("retry_period")
+	if err != nil {
+		return nil, err
+	}
+
+	var argsMapping *bloblang.Executor
+	var keysMapping *bloblang.Executor
+
+	var script string
+	if script, err = conf.FieldString("script"); err != nil {
+		return nil, err
+	}
+
+	redisScript := redis.NewScript(script)
+
+	if argsMapping, err = conf.FieldBloblang("args_mapping"); err != nil {
+		return nil, err
+	}
+
+	if keysMapping, err = conf.FieldBloblang("keys_mapping"); err != nil {
+		return nil, err
+	}
+
+	r := &redisScriptProc{
+		log: res.Logger(),
+
+		script:      redisScript,
+		argsMapping: argsMapping,
+		keysMapping: keysMapping,
+
+		retries:     retries,
+		retryPeriod: retryPeriod,
+		client:      client,
+	}
+
+	return r, nil
+}
+
+func (r *redisScriptProc) exec(ctx context.Context, index int, inBatch service.MessageBatch, msg *service.Message) error {
+	args, err := getArgsMapping(inBatch, index, r.argsMapping)
+
+	if err != nil {
+		return fmt.Errorf("args_mapping failed: %w", err)
+	}
+
+	keys, err := getKeysStrMapping(inBatch, index, r.keysMapping)
+
+	if err != nil {
+		return fmt.Errorf("keys_mapping failed: %w", err)
+	}
+
+	res, err := r.script.Run(ctx, r.client, keys, args...).Result()
+
+	for i := 0; i <= r.retries && err != nil; i++ {
+		r.log.Errorf("script failed: %v", err)
+		<-time.After(r.retryPeriod)
```

Comment:
Wouldn't `time.Sleep()` be more appropriate here?

Reply:
I don't think it matters much in this case.. I used `time.After` because I saw this usage in other files inside benthos codebase.

Reply:
Oh, OK. Just thought it looks better since I only expect to see `time.After()` in `select` statements. Otherwise, there’s no difference.